### PR TITLE
Layoutlm tf

### PR DESCRIPTION
**Pull request by atahmasb**
# What does this PR do?

This PR adds TF version of LayoutLM for issue [(10312)](https://github.com/huggingface/transformers/issues/10312)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

**Comment by LysandreJik** [comment]
Very nice! Can you let us know when you want us to review/give feedback/help? Thanks!

**Comment by atahmasb** [comment]
> Very nice! Can you let us know when you want us to review/give feedback/help? Thanks!

thanks! I need to upload the TF model file to the hub and run another check to make sure it gives the same results as the PT version. I had verified that but I made few changes so I am gonna run the checks one  more time. I'll tag you when it's done

**Comment by atahmasb** [comment]
@LysandreJik I've uploaded TF models to the model hub under :
- atahmasb/tf-layoutlm-base-uncased
- atahmasb/tf-layoutlm-large-uncased

I appreciate if you and the team could take a look at the code and give me feedback.
There are some tests that are failing, I haven't figured the issues out but the code is up for review. Maybe someone could look into the logs and guide me on how to fix them.
Meanwhile I'll see if I can make the tests pass.

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -1335,6 +1344,129 @@
...
+        TF_DPR_CONTEXT_ENCODER_PRETRAINED_MODEL_ARCHIVE_LIST,
+        TF_DPR_QUESTION_ENCODER_PRETRAINED_MODEL_ARCHIVE_LIST,
+        TF_DPR_READER_PRETRAINED_MODEL_ARCHIVE_LIST,
+        TFDPRContextEncoder,
+        TFDPRPretrainedContextEncoder,
+        TFDPRPretrainedQuestionEncoder,
+        TFDPRPretrainedReader,
+        TFDPRQuestionEncoder,
+        TFDPRReader,
+    )
```
Comment (LysandreJik) [inline src/transformers/__init__.py:1356]: Is there a reason you have added all of this here? This should already be covered by the init

File: src/transformers/models/layoutlm/modeling_tf_layoutlm.py [hunk 2]
```diff
@@ -0,0 +1,1202 @@
...
+        inputs_embeds: tf.Tensor = None,
+        training: bool = False,
+    ) -> tf.Tensor:
+        """
+        Applies embedding based on inputs tensor.
+
+        Returns:
+            final_embeddings (:obj:`tf.Tensor`): output embedding tensor.
+        """
+        assert not (input_ids is None and inputs_embeds is None)
```
Comment (LysandreJik) [inline src/transformers/models/layoutlm/modeling_tf_layoutlm.py:146]: Would be nice to add a message for the assertion if it fails.

File: src/transformers/models/layoutlm/modeling_tf_layoutlm.py [hunk 3]
```diff
@@ -0,0 +1,1202 @@
...
+        if bbox is None:
+            bbox = bbox = tf.fill(input_shape + [4], value=0)
+        
+        try:
+            left_position_embeddings = tf.gather(self.x_position_embeddings, bbox[:, :, 0])
+            upper_position_embeddings = tf.gather(self.y_position_embeddings, bbox[:, :, 1])
+            right_position_embeddings = tf.gather(self.x_position_embeddings, bbox[:, :, 2])
+            lower_position_embeddings = tf.gather(self.y_position_embeddings, bbox[:, :, 3])
+        except IndexError as e:
+            raise IndexError("The :obj:`bbox`coordinate values should be within 0-1000 range.") from e
```
Comment (LysandreJik) [inline src/transformers/models/layoutlm/modeling_tf_layoutlm.py:168]: Very nice error message

**Review by LysandreJik — COMMENTED** [review]
Hi @atahmasb, this looks great! I've only left a few comments, everything looks good. I'll do a deeper review once all the tests pass, as things are bound to change until they do, but the idea here is sound!

Do you need any help to make the tests pass?

File: src/transformers/__init__.py (hunk 1 above)
Comment (atahmasb) [reply src/transformers/__init__.py:1356]: I think these lines were added when I rebased, I'll remove these lines.

File: src/transformers/models/layoutlm/modeling_tf_layoutlm.py (hunk 2 above)
Comment (atahmasb) [reply src/transformers/models/layoutlm/modeling_tf_layoutlm.py:146]: Good idea, will add that

**Comment by atahmasb** [comment]
> Hi @atahmasb, this looks great! I've only left a few comments, everything looks good. I'll do a deeper review once all the tests pass, as things are bound to change until they do, but the idea here is sound!
> 
> Do you need any help to make the tests pass?

I am going to try one more time to resolve them today, if I can't then I'll ask for help

**Comment by atahmasb** [comment]
@LysandreJik all tests passed! It's ready for a deeper review please

**Comment by LysandreJik** [comment]
Thanks for letting me know! It seems GitHub botched your rebase, as it is showing 53 files changed. Could you close this PR and open a new one (no need to do anything on your branch) so that we may see the diff better?

Thanks!

**Comment by atahmasb** [comment]
> Thanks for letting me know! It seems GitHub botched your rebase, as it is showing 53 files changed. Could you close this PR and open a new one (no need to do anything on your branch) so that we may see the diff better?
> 
> Thanks!

sure, will do

**Pull request closed without merging**